src/libutil/position.hh

```diff
@@ -63,6 +64,15 @@ struct Pos
     bool operator==(const Pos & rhs) const = default;
     auto operator<=>(const Pos & rhs) const = default;
 
+    std::string getSnippetUpTo(const Pos & end) const;
+
+    /**
+     * Get the SourcePath, if the source was loaded from a file.
+     */
+    inline std::optional<SourcePath> getSourcePath() const {
```

Comment:
likely you don't need `inline` because methods defined in structs are implicitly inlined.